pghoard/rohmu/object_storage/sftp.py

```diff
@@ -0,0 +1,209 @@
+                else:
+                    yield IterKeyItem(type=KEY_TYPE_PREFIX, value=file_key)
+            else:
+                # Don't return files if metadata file is not present; files are written in two phases and
+                # should be considered available only after also metadata has been written
+                if attr.filename + ".metadata" in files:
+                    if with_metadata:
+                        metadata = self.get_metadata_for_key(file_key)
+                    else:
+                        metadata = None
+
+                    last_modified = datetime.datetime.fromtimestamp(attr.st_mtime, tz=datetime.timezone.utc)
+                    yield IterKeyItem(
+                        type=KEY_TYPE_OBJECT,
+                        value={
+                            "name": file_key,
+                            "size": attr.st_size,
+                            "last_modified": last_modified,
+                            "metadata": metadata,
+                        }
+                    )
+                else:
+                    continue
+
+    # can't support remote copy, only remote rename
+    def copy_file(self, *, source_key, destination_key, metadata=None, **_kwargs):
+        raise NotImplementedError
+
+    def delete_key(self, key):
+        target_path = self.format_key_for_backend(key.strip("/"))
+        self.log.info("Removing path: %r", target_path)
+        try:
+            self.client.remove(target_path)
+        except FileNotFoundError as ex:
+            raise FileNotFoundFromStorageError(key) from ex
+
+    def store_file_from_memory(self, key, memstring, metadata=None, cache_control=None, mimetype=None):
+        bio = BytesIO(memstring)
+        try:
+            self.store_file_object(key=key, fd=bio, cache_control=cache_control, metadata=metadata, mimetype=mimetype)
+        except OSError as ex:
+            raise StorageError(key) from ex
+
+    def store_file_from_disk(self, key, filepath, metadata=None, multipart=None, cache_control=None, mimetype=None):
+        with open(filepath, "rb") as fh:
+            self.store_file_object(key=key, fd=fh, cache_control=cache_control, metadata=metadata, mimetype=mimetype)
+
+    # pylint: disable=unused-argument
+    def store_file_object(self, key, fd, *, cache_control=None, metadata=None, mimetype=None, upload_progress_fn=None):
+        target_path = self.format_key_for_backend(key.strip("/"))
+
+        self.log.debug("Store path: %r", target_path)
+
+        def wrapper_upload_progress_fn(bytes_written, total_bytes):  # pylint: disable=unused-argument
+            if upload_progress_fn:
+                upload_progress_fn(bytes_written)
+
+        self.client.makedirs(os.path.dirname(target_path))
+        self.client.putfo(flo=fd, remotepath=target_path, callback=wrapper_upload_progress_fn)
+        self._save_metadata(target_path, metadata)
```

Comment:
Saving metadata first would prevent running into race conditions when listing files where you get back files that have contents but no metadata.

Reply:
The metadata is also the flag file and if a data file has no flag it's not returned in the list thus why it's written only after the data file. I copied this idea from the local provider, this affects some of your comments above too, but when I get to a computer I will review in more detail to ensure that this is working as I intended.